### ISSUE TITLE
[RFC] Swiftgen integration

### ISSUE THREAD

**Issue by fortmarek** [tuist/tuist#1595]
## Need/problem

As @pepibumur has rightly pointed out, since we now derive a new file to easily access resources (even in static frameworks), it'd be great to take even tighter control and generate the namespace for them.

## Motivation

Most teams do not depend on the string representation of images, localized strings etc., but generate a type-safe namespace. Rather than every team having to do this on their own, we can integrate it ourselves.

## Detailed design

As most teams use [swiftgen](https://github.com/SwiftGen/SwiftGen) for the aforementioned purpose, we should not reinvent the wheel and we can use this tool to generate the appropriate files. Since we know the list of resources for every target, we can map this list and check each resource's extension. If we then confirm that it is something that can be generated by swiftgen (eg `.xcassets` will signal images, `.strings` localized string), then we run swiftgen with the appropriate commands. If a resource is not supported by swiftgen - eg. `.txt` - we simply skip it. For this purpose we should be able to use one of the packages that swiftgen defines [here](https://github.com/SwiftGen/SwiftGen/blob/stable/Package.swift). 
The things might get little harder if we want to support regenerating the files on each build with a build phase. I am not sure yet what should be done in the build phase - the most clean solution would probably be to introduce a new command or a new flag to `tuist generate` such as `tuist generate --files-only` (the naming can be improved). We can then create a new build phase that'd call this command. 

Special consideration should be made for static frameworks. For them to work we can define our own custom template which would leverage work done for bringing resources to them in the first place.

Users should also be able to opt-out of this behavior if they want to integrate swiftgen themselves. We can use something similar to `disableAutogeneratedSchemes` in `Config.swift` - eg `disableAutogeneratedNamespace` (feel free to offer suggestions for better naming)

# Drawbacks

- It's another thing we need to take care of, but since almost every team does this, it is IMHO worth it
- We will need to set appropriate defaults for the generated swiftgen files - for example you can pass `--params publicAccess=true` to make the namespace public or `--params enumName="MyEnvironment"` for namespace of a `.plist` file to have a custom name. We could either try to make this modifiable with some parameter in `Config.swift` or, which is something we should try to first, set the default well-enough that no modifications will be necessary
- If teams use swiftgen already, adding the behavior by default would be a breaking change - users will have to opt out

## Alternatives

Not doing this at all and keep the current status quo

# Adoption strategy

It will be a breaking change as mentioned above

# How we teach this

We will need to add a documentation explaining how can users leverage this feature (in this case it is doing nothing but they will have to be aware that the namespace already exists)

We will also need to explain how they can opt-out of this behavior

# Unresolved questions

To summarize unresolved questions from this issue:
- Should we generate a build phase to regenerate files on incremental build and if so, how should that be implemented?
- What should be the default parameters for a template, so most users will not have to alter it?
- How should be users be able to opt-out of the behavior (and what should be the name if we go for `Config.swift`)?

**Comment by pepibumur** [comment]
Hey @fortmarek thanks for the write-up. One thing that I'd suggest changing is how we integrate with SwiftGen. If we rely on SwiftGen being installed, for example via Homebrew, it's possible that developers end up getting different results because they have different versions of SwiftGen. This is frustration for users, and a false impression that the issue is Tuist and not SwiftGen.

My suggestion is to rather link SwiftGen's core library. The project is a Swift Package so we can define a dependency with it and consume the interfaces directly. The drawback is that we'd need to keep up with updates, but the experience that we'd provide to developers is much better.

I think we could provide our own templates that make use of the auto-generated synthesized accessors. We'd need to adjust the installation scrips to vendor those templates.

**Comment by pepibumur** [comment]
Hey @fortmarek thanks for the write-up. One thing that I'd suggest changing is how we integrate with SwiftGen. If we rely on SwiftGen being installed, for example via Homebrew, it's possible that developers end up getting different results because they have different versions of SwiftGen. This is frustration for users, and a false impression that the issue is Tuist and not SwiftGen.

My suggestion is to rather link SwiftGen's core library. The project is a Swift Package so we can define a dependency with it and consume the interfaces directly. The drawback is that we'd need to keep up with updates, but the experience that we'd provide to developers is much better.

I think we could provide our own templates that make use of the auto-generated synthesized accessors. We'd need to adjust the installation scrips to vendor those templates.

**Comment by fortmarek** [comment]
Thanks for the reply @pepibumur!

> My suggestion is to rather link SwiftGen's core library.

That's what I wanted to do, but I have not made that clear in the initial post, so thanks for pointing that out!

> think we could provide our own templates that make use of the auto-generated synthesized accessors

I agree - we can do it in similar fashion as we bundle templates for `scaffold` command 🙂 

I'll make sure to link here a draft PR (once there is one 😄) and we can discuss further details there

**Comment by pepibumur** [comment]
Awesome! The plan looks great. It's gonna be so easy to adopt ❤️
